### DISCUSSION TITLE
Run script on android 7 and lower versions with frida gadget

### DISCUSSION THREAD

**Discussion by AlephBaa** [vfsfitvnm/frida-il2cpp-bridge#263] [Q&A] [answered]
Hi
I failed to run the script produced by the library on Android 7 and lower versions with frida gadget. The app runs normally like there is no scripts. The script and frida gadget does their jobs good on Android +8. Is it compatibility problem? I didn't find any discussion about this nor in ReadMe and I don't have android 7 or lower device unfortunately and can't check logs.
I test a basic script (without using il2cpp-bridge) and that works.
any suggestion?

**Comment by Flechaa** [comment, marked as answer]
> Hi I failed to run the script produced by the library on Android 7 and lower versions with frida gadget. The app runs normally like there is no scripts. The script and frida gadget does their jobs good on Android +8. Is it compatibility problem? I didn't find any discussion about this nor in ReadMe and I don't have android 7 or lower device unfortunately and can't check logs. I test a basic script (without using il2cpp-bridge) and that works. any suggestion?

So you are saying Il2Cpp.perform() is never performed?

**Reply by AlephBaa** [reply]
As i said i don't have any android 7 and lower version to test. But the modifications of the application exist on android 8, 9, 10 ... and doesn't exist on android 5,6,7
I can say yes. That doesn't perform.

**Reply by Flechaa** [reply]
I do have an android 7 device and that also happens, i can't confirm if we are talking about the same thing, i figured out that if you add a timeout to Il2Cpp.perform() and retry it after that timeout the callback will be called as intended however it's more a workaround than a fix.

**Reply by AlephBaa** [reply]
Thanks. Can you provide me the general structure of the code you used? I'll test it

**Reply by Flechaa** [reply]
Sure, I used something like this.
```
import "frida-il2cpp-bridge";

class Injector {
    protected static TimeoutDelay = 5000;
    protected static timeout: NodeJS.Timeout | null = null;
    public static CSharpAssembly: Il2Cpp.Image;

    static {
        Injector.timeout = setInterval(Injector.setup, Injector.TimeoutDelay);
        Injector.setup();
    }

    protected static setup() {
        if (!Injector.timeout) return;
        Il2Cpp.perform(Injector.tryAttach);
    }

    protected static tryAttach() {
        clearInterval(Injector.timeout as NodeJS.Timeout);
        if (!Injector.timeout) return;
        Injector.awaitIl2Cpp();
        Injector.timeout = null;
    }

    private static awaitIl2Cpp() {
        const interval = setInterval(() => {
            if (!Module.findBaseAddress("libil2cpp.so")) return;
            try {
                Injector.attach();
            } catch (err) {
                console.error(err);
            }
            clearInterval(interval);
        }, 1000);
    }


    private static loadImages() {
        Injector.CSharpAssembly = Il2Cpp.Domain.assembly("Assembly-CSharp").image;
    }

    private static attach() {
        Injector.loadImages();
	console.log("Il2Cpp is now attached");
    }
}
```

**Reply by AlephBaa** [reply]
Error: could not parse './_agent.js' line 1: invalid property name
    at <anonymous> (/frida/repl-2.js:1)

_agent.js had generated with with webpack.

**Reply by Flechaa** [reply]
I used esbuild to compile this

**Reply by HorridModz** [reply]
Thank you SO MUCH! After searching for hours, I found your code, and it worked. It wouldn't compile at first, but here's what worked for me:

index.ts:
```ts
import "frida-il2cpp-bridge";

class Injector {
    protected static TimeoutDelay = 5000;
    protected static timeout: NodeJS.Timeout | null = null;
    public static CSharpAssembly: Il2Cpp.Image;

    static initialize() {
        Injector.timeout = setInterval(Injector.setup, Injector.TimeoutDelay);
        Injector.setup();
    }

    protected static setup() {
        if (!Injector.timeout) return;
        Il2Cpp.perform(Injector.tryAttach);
    }

    protected static tryAttach() {
        clearInterval(Injector.timeout as NodeJS.Timeout);
        if (!Injector.timeout) return;
        Injector.awaitIl2Cpp();
        Injector.timeout = null;
    }

    private static awaitIl2Cpp() {
        const interval = setInterval(() => {
            if (!Module.findBaseAddress("libil2cpp.so")) return;
            try {
                Injector.attach();
            } catch (err) {
                console.error(err);
            }
            clearInterval(interval);
        }, 1000);
    }

    private static loadImages() {
        Injector.CSharpAssembly = Il2Cpp.domain.assembly("Assembly-CSharp").image;
    }

    private static attach() {
        Injector.loadImages();
    }
}

// Call the initialize method
Injector.initialize();

Il2Cpp.perform(() => {
    console.log("Il2Cpp is now attached");
})
```

The static block was causing issues, so I moved it into an `initialize()` method.

I also needed some changes in `package.json` (to the build, attach, and spawn scripts):"

```json
{
    "scripts": {
      "build": "npm run build-ts && npm run build-es",
      "build-ts": "tsc",
      "build-es": "esbuild index.ts --bundle --outfile=_.js --platform=node --target=node14",
      "attach": "run() { frida -U \"$1\" -l _.js --runtime=v8; }; run",
      "spawn": "run() { frida -U -f \"$1\" -l _.js --no-pause --runtime=v8; }; run",
      "app0-spawn": "npm run spawn com.example.application0",
      "app1": "npm run \"Application1 Name\"",
      "app1-spawn": "npm run spawn com.example.application1"
    },
    "devDependencies": {
      "@types/frida-gum": "^18.3.1",
      "esbuild": "^0.14.15",
      "frida-il2cpp-bridge": "*",
      "typescript": "^4.5.5"
    }
  }
  ```
  
  Of course, I also had to install esbuild:
  
  ```sh
  npm install -g esbuild
  ```

**Reply by Flechaa** [reply]
It should already be fixed in the latest version, at least for me.
